wsmux/session.go

```diff
@@ -0,0 +1,351 @@
+	s.streams[id] = str
+
+	if err := s.send(newSynFrame(id)); err != nil {
+		s.nextID -= 2
+		return nil, err
+	}
+	// unlock mutex and wait
+	s.mu.Unlock()
+
+	select {
+	case <-str.accepted:
+		s.mu.Lock()
+		return str, nil
+	case <-s.closed:
+		s.mu.Lock()
+		// state of s.nextID doesn't matter here
+		delete(s.streams, id)
+		return nil, ErrSessionClosed
+	case <-time.After(s.streamAcceptDeadline):
+		s.mu.Lock()
+		// nextID can be cyclically reused, and previous instance
+		// may be in use by a different stream
+		delete(s.streams, id)
+		return nil, ErrAcceptTimeout
+	}
+
+}
+
+// Close closes the current session and underlying connection
+func (s *Session) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Check if channel has been closed
+	if s.IsClosed() {
+		return nil
+	}
+
+	var err error
+	if s.closeConn {
+		err = s.conn.Close()
+	}
+
+	for _, v := range s.streams {
+		_ = v.Close()
+	}
+	s.streams = nil
+
+	close(s.closed)
+	return err
+}
+
+// Addr used for implementing net.Listener
+func (s *Session) Addr() net.Addr {
+	return s.conn.LocalAddr()
+}
+
+func (s *Session) closeHandler(code int, text string) error {
+	s.logger.Printf("ws conn closed: code %d : %s", code, text)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.closeConn = false
+	return s.Close()
+}
+
+func (s *Session) recvLoop() {
+	for {
+		select {
+		case <-s.closed:
+			return
+		default:
+		}
+
+		t, msg, err := s.conn.ReadMessage()
+		if err != nil {
+			_ = s.abort(err)
+			break
+		}
+		if t != websocket.BinaryMessage {
+			s.logger.Print("did not receive binary message")
+			continue
+		}
+
+		if len(msg) < 5 {
+			s.logger.Print(errMalformedHeader)
+			_ = s.abort(errMalformedHeader)
+		}
+
+		h := header(msg[:5])
+		msg = msg[5:]
+
+		id, msgType := h.id(), h.msg()
+
+		switch msgType {
+		//Used for creating a new stream
+		case msgSYN:
+			s.mu.Lock()
+			if _, ok := s.streams[id]; ok {
+				s.logger.Printf("received duplicate syn frame for stream %d", id)
+				s.mu.Unlock()
+				break
+			}
+
+			str := newStream(id, s)
+			// no point in locking here
+			str.accept(DefaultCapacity)
+
+			s.streams[id] = str
+			if err := s.send(newAckFrame(id, uint32(DefaultCapacity))); err != nil {
+				s.logger.Printf("error accepting stream %d", id)
+				_ = s.Close()
+				return
+			}
+			s.asyncPushStream(str)
+			s.mu.Unlock()
+
+		//received data
+		case msgDAT:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			if !ok {
+				s.logger.Printf("received data frame for unknown stream %d", id)
+				break
+			}
+			str.push(msg)
+			s.logger.Printf("received DAT frame on stream %d: %v", id, bytes.NewBuffer(msg))
+
+		//received ack frame
+		case msgACK:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			s.logger.Printf("received ack frame")
+			if !ok {
+				s.logger.Printf("received ack frame for unknown stream %d", id)
+				break
+			}
+
+			read := binary.LittleEndian.Uint32(msg)
+			select {
+			case <-str.accepted:
```

Comment:
Actually, since `stream` isn't exported... let's start methods and properties used by `Session` with capital...
`stream.OnData(b []byte)` that makes it more clear that the method is not just an internal helper function...